std/functional.d

```diff
@@ -1467,3 +1467,184 @@ template forward(args...)
+        while (newLen > retBuff.length)
+            retBuff.length *= 2;
+
+        ret = retBuff[0 .. newLen];
+        ret[($ - str.length) .. $] = str;
+    }
+
+    // Ideally this wrapper should never appear explicitly in the emitted binary.
+    append("pragma(inline, true) ");
+
+    if (fas & FA.property)
+        append("@property ");
+    if (fas & FA.ref_)
+        append("ref ");
+
+    append("auto ");
+    append(arvName);
+    append("()(");
+
+    static if (names.length > 0)
+    {
+        alias types = traits.Parameters!_func;
+        alias stcs  = traits.ParameterStorageClassTuple!_func;
+        alias dflts = traits.ParameterDefaults!_func;
+
+        foreach (p, name; names)
+        {
+            // storage classes
+            if ((stcs[p] & STC.lazy_) != 0)
+                append("lazy ");
+            if ((stcs[p] & STC.scope_) != 0)
+                append("scope ");
+            if ((stcs[p] & STC.out_) != 0)
+                append("out ");
+
+            if ((stcs[p] & STC.return_) != 0)
+                append("return ref ");
+            else
+            if ((stcs[p] & STC.ref_) != 0)
+                append("auto ref ");
+
+            // type
+            append(types[p].stringof);
+            append(" ");
+
+            // identifier
+            if (name.length > 0)
+                append(name);
+            else
+            {
+                append("a");
+                append(p.stringof);
+            }
+
+            // default value
+            static if (! is(dflts[p] == void))
+            {
+                append(" = ");
+                append(dflts[p].stringof);
+            }
+
+            if ((p + 1) < names.length)
+                append(", ");
+        }
+    }
+    append(") ");
+
+    if (fas & FA.const_)
+        append("const ");
+    if (fas & FA.shared_)
+        append("shared ");
+
+    append("{ return _func(");
+    static if (names.length > 0)
+    {
+        foreach (p, name; names)
+        {
+            if (name.length > 0)
+                append(name);
+            else
+            {
+                append("a");
+                append(p.stringof);
+            }
+
+            if ((p + 1) < names.length)
+                append(", ");
+        }
+    }
+    append("); }");
+    return cast(string)ret;
+}
+
+///
+unittest {
+    int addLVals(ref int a, ref int b)
+    {
+        const ret = a + b;
+        a = 0;
+        return ret;
+    }
+
+    // This will not compile, because 1 and 2 are rvalues:
+    static assert(! __traits(compiles, addLVals(1, 2)));
+
+    // This *does* compile:
+    mixin acceptRVals!("addAny", addLVals);
+    assert(addAny(1, 2) == 3);
+
+    // lvalues are still passed by reference:
+    int x = 5, y = 3;
+    assert(addAny(x, y) == 8);
+    assert(x == 0);
+
+    // The wrapped function can be anonymous:
+    mixin acceptRVals!("mul", function(ref const(int) a, ref const(int) b) {
```

Comment:
For readability's sake, could you please break these up into multiple ddoc-ed unit tests.